tls.c

```diff
@@ -81,54 +83,85 @@ ssize_t ssl_write(conn *c, void *buf, size_t count) {
     return SSL_write(c->ssl, buf, count);
 }
 
+/*
+ * Prints an SSL error into the buff, if there's any.
+ */
+static void print_ssl_error(char *buff, size_t len) {
+    unsigned long err;
+    if ((err = ERR_get_error()) != 0) {
+        ERR_error_string_n(err, buff, len);
+    }
+}
+
 /*
  * Loads server certificates to the SSL context and validate them.
  * @return whether certificates are successfully loaded and verified or not.
  * @param error_msg contains the error when unsuccessful.
  */
 static bool load_server_certificates(char **errmsg) {
-    bool success = true;
-    char *error_msg = malloc(MAXPATHLEN + MAX_ERROR_MSG_SIZE);
-    size_t errmax = MAXPATHLEN + MAX_ERROR_MSG_SIZE - 1;
+    bool success = false;
+
+    char *error_msg = malloc(MAXPATHLEN + ERROR_MSG_SIZE +
+        SSL_ERROR_MSG_SIZE);
+    size_t errmax = MAXPATHLEN + ERROR_MSG_SIZE + SSL_ERROR_MSG_SIZE - 1;
+
     if (error_msg == NULL) {
         *errmsg = NULL;
         return false;
     }
+
     if (settings.ssl_ctx == NULL) {
         snprintf(error_msg, errmax, "Error TLS not enabled\r\n");
         *errmsg = error_msg;
         return false;
     }
+
+    char *ssl_err_msg = malloc(SSL_ERROR_MSG_SIZE);
+    if (ssl_err_msg == NULL) {
+        *errmsg = NULL;
+        return false;
+    }
+    bzero(ssl_err_msg, SSL_ERROR_MSG_SIZE);
+
     SSL_LOCK();
     if (!SSL_CTX_use_certificate_chain_file(settings.ssl_ctx,
         settings.ssl_chain_cert)) {
-        snprintf(error_msg, errmax, "Error loading the certificate chain: %s\r\n",
-            settings.ssl_chain_cert);
-        success = false;
+        print_ssl_error(ssl_err_msg, SSL_ERROR_MSG_SIZE);
+        snprintf(error_msg, errmax, "Error loading the certificate chain: "
+            "%s : %s\r\n", settings.ssl_chain_cert, ssl_err_msg);
     } else if (!SSL_CTX_use_PrivateKey_file(settings.ssl_ctx, settings.ssl_key,
                                         settings.ssl_keyformat)) {
-        snprintf(error_msg, errmax, "Error loading the key: %s\r\n", settings.ssl_key);
-        success = false;
+        print_ssl_error(ssl_err_msg, SSL_ERROR_MSG_SIZE);
+        snprintf(error_msg, errmax, "Error loading the key: %s : %s\r\n",
+            settings.ssl_key, ssl_err_msg);
     } else if (!SSL_CTX_check_private_key(settings.ssl_ctx)) {
-        snprintf(error_msg, errmax, "Error validating the certificate\r\n");
-        success = false;
+        print_ssl_error(ssl_err_msg, SSL_ERROR_MSG_SIZE);
+        snprintf(error_msg, errmax, "Error validating the certificate: %s\r\n",
```

Comment:
Here and above/below, should check the result of snprintf() to see if the call was truncated; since that will snip the `\r\n` and cause a protocol hang if we somehow get a huge error from OpenSSL.